### ISSUE TITLE
Implement `dash::mutex` for mutual exclusion 

### ISSUE THREAD

**Issue by fmoessbauer** [dash-project/dash#327]
Mutex similar to `std::mutex` based on the synchronization interface of DART.

**Comment by fuchsto** [comment]
Is there a specific use case where a mutex is required or is this a general feature request?

**Comment by fmoessbauer** [comment]
It's both. In [Johnson, Coarray C++, 6.3](http://www.pgas2013.org.uk/sites/default/files/finalpapers/Day2/R4/2_paper6.pdf) he proposes a mutex which can be placed in his coarray and behaves similar to a `std::mutex`.

My idea was to implement it roughly the same way as `dash::Atomic` by specializing `dash::GobRef<dash::Mutex>`. Together with #326 this would enable the following use case:

```cpp
dash::CoArray<dash::Mutex> m;
m(i).lock()
// access data on image i
m(i).unlock();
```

**Comment by fuchsto** [comment]
No, please don't. The `GlobRef<Atomic>` is a **phantom** and never explicitly defined by / for application developers.
Specializing `GlobRef` for a visible / explicit type is like `std::vector<bool>`. Looks convenient, is plain hell.

**Comment by fmoessbauer** [comment]
This is a dilemma: The CoArray should behave almost exactly like the stored value, so for mutexes it's like the following:

```cpp
std::vector<std::mutex> mxstd(10);
dash::CoArray<dash::mutex> mxdash;

mxstd[0].lock();   // OK
mxdash(0).lock(); // Not OK ??
```

But after sleeping a night over this, you are right that this is not a good decision: I could make it possible to mimic the interface, but almost all semantics will change as a `dash::mutex` requires a copy per unit, while `std::mutex` requires exactly one copy for all threads. So what about introducing a `dash::CoMutex` which can be accessed similar to the `dash::CoArray<T>` but as it is its own container, this is well defined:

```cpp
dash::CoMutex comx;
comx(0).lock();
comx(0).unlock();
```

**Comment by fmoessbauer** [comment]
Implemented in #329
